lib/src/database/indexeddb_box.dart

```diff
@@ -0,0 +1,194 @@
+import 'dart:async';
+import 'dart:html';
+import 'dart:indexed_db';
+
+/// Key-Value store abstraction over IndexedDB so that the sdk database can use
+/// a single interface for all platforms. API is inspired by Hive.
+class BoxCollection {
+  final Database _db;
+  final Set<String> boxNames;
+
+  BoxCollection(this._db, this.boxNames);
+
+  static Future<BoxCollection> open(
+    String name,
+    Set<String> boxNames, {
+    int version = 1,
+    Object? sqfliteDatabase,
+    IdbFactory? idbFactory,
+  }) async {
+    idbFactory ??= window.indexedDB!;
+    final db = await idbFactory.open(name, version: 1,
+        onUpgradeNeeded: (VersionChangeEvent event) {
+      final db = event.target.result;
+      for (final name in boxNames) {
+        db.createObjectStore(name, autoIncrement: true);
+      }
+    });
+    return BoxCollection(db, boxNames);
+  }
+
+  Box<V> openBox<V>(String name) {
+    if (!boxNames.contains(name)) {
+      throw ('Box with name $name is not in the known box names of this collection.');
+    }
+    return Box<V>(name, this);
+  }
+
+  List<Future<void> Function(Transaction txn)>? _txnCache;
+
+  Future<void> transaction(
+    Future<void> Function() action, {
+    List<String>? boxNames,
+    bool readOnly = false,
+  }) async {
+    boxNames ??= _db.objectStoreNames!.toList();
+    _txnCache = [];
+    await action();
+    final cache = List<Future<void> Function(Transaction txn)>.from(_txnCache!);
+    _txnCache = null;
+    if (cache.isEmpty) return;
+    final txn = _db.transaction(boxNames, readOnly ? 'readonly' : 'readwrite');
+    for (final fun in cache) {
+      unawaited(fun(txn));
```

Comment:
Why are these unawaited? That doesn't seem to make sense with marking the txn as completed? A comment would be appreciated :D

Reply:
Added a comment